pkg/client/typed_client.go

```diff
@@ -143,6 +144,68 @@ func (c *typedClient) Get(ctx context.Context, key ObjectKey, obj Object) error
 		Name(key.Name).Do(ctx).Into(obj)
 }
 
+func (c *typedClient) ListPages(ctx context.Context, obj ObjectList,
+	callback func(obj ObjectList) error, opts ...ListOption) error {
+	r, err := c.cache.getResource(obj)
+	if err != nil {
+		return err
+	}
+	listOpts := ListOptions{}
+	listOpts.ApplyOptions(opts)
+
+	// Fetch items at chunks of one hundred if not specified differently.
+	if listOpts.Limit == 0 {
+		Limit(DefaultPageLimit).ApplyToList(&listOpts)
+	}
+
+	// Retrieve initial chunck of data.
+	var allItems []runtime.Object
+	var interimResult ObjectList
+	err = r.Get().
+		NamespaceIfScoped(listOpts.Namespace, r.isNamespaced()).
+		Resource(r.resource()).
+		VersionedParams(listOpts.AsListOptions(), c.paramCodec).
+		Do(ctx).Into(interimResult)
+	if err != nil {
+		return err
+	}
+
+	if err := callback(interimResult); err != nil {
+		return err
+	}
+
+	items, err := apimeta.ExtractList(interimResult)
+	if err != nil {
+		return err
+	}
+	allItems = append(allItems, items...)
+
+	// Continue while there are more chunks.
+	for interimResult.GetContinue() != "" {
+		Continue(interimResult.GetContinue()).ApplyToList(&listOpts)
+		err = r.Get().
+			NamespaceIfScoped(listOpts.Namespace, r.isNamespaced()).
+			Resource(r.resource()).
+			VersionedParams(listOpts.AsListOptions(), c.paramCodec).
+			Do(ctx).Into(interimResult)
+		if err != nil {
+			return err
+		}
+
+		if err := callback(interimResult); err != nil {
+			return err
+		}
+
+		items, err = apimeta.ExtractList(interimResult)
+		if err != nil {
+			return err
+		}
+		allItems = append(allItems, items...)
```

Comment:
We don't want both a callback function and a huge list return, we should pick one pattern or the other (or move them to separate functions/options). The reason to use a visitor callback function is to avoid keeping the whole list in memory. I think it would be more Go-like to just do the callback style probably, `ExtractList()` requires a lot of runtime `reflect` magic since there is no unified API for accessing list items. But having both available could be convenient.

Reply:
@vincepri Do you have opinions on which API to support (or both)?

Reply:
Gotcha. :+1:

Reply:
That is she can pass something like this:
```
out := &corev1.Pods{}
allItems := []runtime.Object
cl.ListPages(ctx, out,  func(obj client.ObjectList) error {
    items, err := apimeta.ExtractList(interimResult)
    if err != nil {
        return err
    }
    allItems = append(allItems, items...)
    return nil
})
```

Reply:
Indeed, or more likely cast it to the correct type and skip needing ExtractList :)